File: lib/sql/entries.go
package sql

import (
	"sort"

	"github.com/TheOpenDictionary/odict/lib/types"
	"github.com/TheOpenDictionary/odict/lib/utils"
	"github.com/bokwoon95/sq"
)

type ENTRIES struct {
	sq.TableStruct
	ID            sq.UUIDField `ddl:"primarykey"`
	CREATED_AT    sq.TimeField `ddl:"default=CURRENT_TIMESTAMP"`
	UPDATED_AT    sq.TimeField
	TERM          sq.StringField `ddl:"notnull"`
	DICTIONARY_ID sq.UUIDField   `ddl:"references=dictionaries.id"`
}

// Insert entries w/ relation to dictionary
func insertEntries(builder *SQLBuilder, dictionaryID string, entryMap types.KVMap[string, types.Entry]) {
	entries := entryMap.Values()

	sort.Slice(entries, func(i, j int) bool {
		return entries[i].Term < entries[j].Term
	})

	for _, entry := range entries {
		e := sq.New[ENTRIES]("")
		id := utils.CreateUUID()
		term := entry.Term

		builder.AddCommand(
			sq.
				InsertInto(e).
				Columns(e.ID, e.TERM, e.DICTIONARY_ID).
				Values(sq.Literal(id), sq.Literal(term), sq.Literal(dictionaryID)),
		)

		insertEtymologies(builder, id, entry.Etymologies)
	}
}

File: lib/src/sql/entries.rs
use std::error::Error;

use sea_query::{ColumnDef, ForeignKey, ForeignKeyAction, Iden, Query, Table};

use crate::{Entry, ID};

use super::{dictionaries::Dictionaries, etymologies::insert_etymology, utils::SQLBuilder};

#[derive(Iden)]
pub enum Entries {
    Table,
    ID,
    Term,
    DictionaryID,
}

pub fn create_entries(builder: &mut SQLBuilder) {
    builder.add_table(
        Table::create()
            .table(Entries::Table)
            .if_not_exists()
            .col(ColumnDef::new(Entries::ID).uuid().not_null().primary_key())
            .col(ColumnDef::new(Entries::Term).string().not_null())
            .col(ColumnDef::new(Entries::DictionaryID).uuid().not_null())
            .foreign_key(
                ForeignKey::create()
                    .from(Entries::Table, Entries::DictionaryID)
                    .to(Dictionaries::Table, Dictionaries::ID)
                    .on_delete(ForeignKeyAction::Cascade)
                    .on_update(ForeignKeyAction::Cascade),
            ),
    )
}

pub fn insert_entry(
    builder: &mut SQLBuilder,
    dictionary_id: &str,
    entry: &Entry,
) -> crate::Result<()> {
    let id = ID::default();

    builder.add_insert(
        Query::insert()
            .into_table(Entries::Table)
            .columns([Entries::ID, Entries::Term, Entries::DictionaryID])
            .values([
                id.as_str().into(),
                entry.term.as_str().into(),
                dictionary_id.into(),
            ])?,
    );

    for ety in &entry.etymologies {
        insert_etymology(builder, id.as_str(), ety)?;
    }

    Ok(())
}
